src/k2/persistence/logStream/LogStream.cpp

```diff
@@ -0,0 +1,547 @@
+    }
+    return seastar::when_all_succeed(createFutures.begin(), createFutures.end())
+    .then([this] (auto&& responses){
+        for (auto& response: responses){
+            auto& [status, plogId] = response;
+            if (!status.is2xxOK()){
+                throw k2::dto::PlogCreateError("unable to create plog for Logstream Base");
+            }
+            _preallocatedPlogPool.push_back(std::move(plogId));
+        }
+        return seastar::make_ready_future<>();
+    });
+}
+
+
+seastar::future<>
+LogStreamBase::_activeAndPersistTheFirstPlog(){
+    String plogId = _preallocatedPlogPool.back();
+    _preallocatedPlogPool.pop_back();
+    PlogInfo info{.currentOffset=0, .sealed=false, .next_plogId=""};
+    _first_plogId = plogId;
+    _current_plogId = plogId;
+    _usedPlogInfo[std::move(plogId)] = std::move(info);
+    
+    // persist this used plog info
+    return _addNewPlog(0, _current_plogId)
+    .then([this] (auto&& response){
+        if (!response.is2xxOK()) {
+            throw k2::dto::LogStreamBasePersistError("unable to persist metadata");
+        }
+        _create = true;
+        return seastar::make_ready_future<>();
+    });
+}
+seastar::future<std::tuple<Status, String, uint32_t> > 
+LogStreamBase::append(Payload payload, String plogId, uint32_t offset){
+    if (!_create){
+        return seastar::make_ready_future<std::tuple<Status, String, uint32_t> >(std::tuple<Status, String, uint32_t>(Statuses::S404_Not_Found("LogStreamBase does not created"), "", 0));
+    }
+    if (_current_plogId != plogId || _usedPlogInfo[_current_plogId].sealed || _usedPlogInfo[_current_plogId].currentOffset != offset){
+        return seastar::make_ready_future<std::tuple<Status, String, uint32_t> >(std::tuple<Status, String, uint32_t>(Statuses::S403_Forbidden("LogStreamBase append request information inconsistent"), "", 0));
+    }
+    return append(std::move(payload));
+};
+
+
+seastar::future<std::tuple<Status, String, uint32_t> > 
+LogStreamBase::append(Payload payload){
+    if (!_create){
+        return seastar::make_ready_future<std::tuple<Status, String, uint32_t> >(std::tuple<Status, String, uint32_t>(Statuses::S404_Not_Found("LogStreamBase does not created"), "", 0));
+    }
+    String plogId = _current_plogId;
+
+    if (_usedPlogInfo[plogId].currentOffset + payload.getSize() > PLOG_MAX_SIZE){
+        // switch to a new plog
+        return _switchPlogAndAppend(std::move(payload));
+    }
+    else{
+        uint32_t current_offset = _usedPlogInfo[plogId].currentOffset;
+        _usedPlogInfo[plogId].currentOffset += payload.getSize();
+        uint32_t expect_appended_offset = _usedPlogInfo[plogId].currentOffset;
+        return _client.append(plogId, current_offset, std::move(payload)).
+        then([this, plogId, expect_appended_offset] (auto&& response){
+            auto& [status, appended_offset] = response;
+
+            if (!status.is2xxOK() || expect_appended_offset != appended_offset) {
```

Comment:
Do you mean we need to cache all the pending payloads and call _switchPlogAndAppend when we meet a failure?

Reply:
changed the plogappendresponse. if the append failures, it will return the received payload back to the logstream, and the logstream will call switchandandappendplog